File: src/main/java/com/ra/base_spring_boot/controller/NewsController.java
package com.ra.base_spring_boot.controller;

import com.ra.base_spring_boot.dto.ResponseWrapper;
import com.ra.base_spring_boot.dto.req.FormNew;
import com.ra.base_spring_boot.dto.resp.NewResponse;
import com.ra.base_spring_boot.dto.resp.PageResponse;
import com.ra.base_spring_boot.services.INewsService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/admin/")
@RequiredArgsConstructor
@Tag(name = "News Management", description = "APIs for managing news (Admin only)")
public class NewsController {

    @Autowired
    private INewsService newService;

    @Operation(summary = "Get all news", description = "Retrieve a paginated list of news with optional search")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "Successful retrieval",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "400", description = "Invalid pagination parameters"),
            @ApiResponse(responseCode = "403", description = "Unauthorized access")
    })
    @SecurityRequirement(name = "bearerAuth")
    @GetMapping("/news")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> getAllNews(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int pageSize,
            @RequestParam(defaultValue = "name") String sortField,
            @RequestParam(defaultValue = "asc") String sortOrder,
            @RequestParam(defaultValue = "") String search) {
        PageResponse<NewResponse> newResponse = newService.getAllNews(page, pageSize, sortField, sortOrder, search);
        return ResponseEntity.ok().body(
                ResponseWrapper.builder()
                        .status(HttpStatus.OK)
                        .code(200)
                        .data(newResponse)
                        .build()
        );
    }

    @Operation(summary = "Add a new news item", description = "Create a new news item")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "News created successfully",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "400", description = "Invalid input data"),
            @ApiResponse(responseCode = "403", description = "Unauthorized access")
    })
    @SecurityRequirement(name = "bearerAuth")
    @PostMapping("/news/create")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> addFestival(@Valid @ModelAttribute FormNew formNew) {
        newService.createNew(formNew);
        return ResponseEntity.ok().body(
                ResponseWrapper.builder()
                        .status(HttpStatus.CREATED)
                        .code(201)
                        .data("News created successfully")
                        .build()
        );
    }

    @Operation(summary = "Update a news item", description = "Update an existing news item by ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", description = "News updated successfully",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "400", description = "Invalid input data"),
            @ApiResponse(responseCode = "403", description = "Unauthorized access"),
            @ApiResponse(responseCode = "404", description = "News not found")
    })
    @SecurityRequirement(name = "bearerAuth")
    @PutMapping("/news/update/{id}")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> updateNews(@Valid @PathVariable Long id, @ModelAttribute FormNew formNew) {
        newService.updateNew(id, formNew);
        return ResponseEntity.ok().body(
                ResponseWrapper.builder()
                        .status(HttpStatus.OK)
                        .code(200)
                        .data("News updated successfully")
                        .build()
        );
    }

    @Operation(summary = "Delete a news item", description = "Delete a news item by ID")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "204", description = "News deleted successfully",
                    content = @Content(mediaType = "application/json",
                            schema = @Schema(implementation = ResponseWrapper.class))),
            @ApiResponse(responseCode = "404", description = "News not found"),
            @ApiResponse(responseCode = "403", description = "Unauthorized access")
    })
    @SecurityRequirement(name = "bearerAuth")
    @DeleteMapping("/news/delete/{id}")
    @PreAuthorize("hasAuthority('ROLE_ADMIN')")
    public ResponseEntity<?> deleteNews(@PathVariable Long id) {
        newService.deleteNew(id);
        return ResponseEntity.ok().body(
                ResponseWrapper.builder()
                        .status(HttpStatus.NO_CONTENT)
                        .code(204)
                        .data("News deleted successfully")
                        .build()
        );

    }
}

File: src/controllers/admin/newsController.js
import newsService from '../../services/newsService.js';

class AdminNewsController {
  // Get all news with pagination (Admin API)
  async getAllNews(req, res) {
    try {
      const {
        page = 0,
        pageSize = 10,
        sortField = 'title',
        sortOrder = 'asc',
        search = ''
      } = req.query;
      
      const result = await newsService.getAllNews(
        parseInt(page),
        parseInt(pageSize),
        sortField,
        sortOrder,
        search
      );
      
      res.status(200).json({
        status: 200,
        code: 200,
        data: result
      });
    } catch (error) {
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }

  // Create new news (Admin API)
  async createNews(req, res) {
    try {
      const { title, content } = req.body;
      const imageFile = req.file; // From multer middleware
      
      const news = await newsService.createNews(
        { title, content },
        imageFile
      );
      
      res.status(201).json({
        status: 201,
        code: 201,
        data: 'News created successfully'
      });
    } catch (error) {
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }

  // Update news (Admin API)
  async updateNews(req, res) {
    try {
      const { id } = req.params;
      const { title, content } = req.body;
      const imageFile = req.file; // From multer middleware
      
      await newsService.updateNews(
        id,
        { title, content },
        imageFile
      );
      
      res.status(200).json({
        status: 200,
        code: 200,
        data: 'News updated successfully'
      });
    } catch (error) {
      if (error.message === 'News not found') {
        return res.status(404).json({
          status: 404,
          code: 404,
          message: 'News not found'
        });
      }
      
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }

  // Delete news (Admin API)
  async deleteNews(req, res) {
    try {
      const { id } = req.params;
      
      await newsService.deleteNews(id);
      
      res.status(204).json({
        status: 204,
        code: 204,
        message: 'News deleted successfully'
      });
    } catch (error) {
      if (error.message === 'News not found') {
        return res.status(404).json({
          status: 404,
          code: 404,
          message: 'News not found'
        });
      }
      
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }

  // Get news by ID (Admin API)
  async getNewsById(req, res) {
    try {
      const { id } = req.params;
      
      const news = await newsService.getNewsById(id);
      
      res.status(200).json({
        status: 200,
        code: 200,
        data: news
      });
    } catch (error) {
      if (error.message === 'News not found') {
        return res.status(404).json({
          status: 404,
          code: 404,
          message: 'News not found'
        });
      }
      
      res.status(500).json({
        status: 500,
        code: 500,
        message: error.message
      });
    }
  }
}

export default new AdminNewsController();
